File: solutions/0100-same-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // detailedBfsHelper - A BFS Helper that prints "ALL" nodes even the non-existant ones.
    // Null nodes are appended in traversal as "L," or "R," signifying left-null or right-null node.
    void detailedBfsHelper(TreeNode* root, std::string& traversal, uint8_t height) {
        if (!root->left && !root->right) {
            traversal.append(std::to_string(root->val));
            return;
        }
        std::queue<TreeNode*> q;
        q.push(root);
        q.push(nullptr); // Signifies end of level traversal
        height--;
        while(!q.empty()) {
            TreeNode* cur = q.front(); q.pop();
            if (!cur) {
                if (q.empty()) {
                    // All nodes processed
                    break;
                }
                // decrement height
                height--;
                q.push(nullptr);
                continue;
            } else {
                if (cur->val > -10001 && cur->val < 10001) {
                    traversal.append(std::to_string(cur->val));
                    traversal.append(",");
                } else if (cur->val == -10001) {
                    traversal.append("L,");
                } else if (cur->val == 10001) {
                    traversal.append("R,");
                }
            }
            if (cur->left) {
                q.push(cur->left);
            } else {
                // Leaf node, don't add null nodes for height = 1
                if (height > 0) {
                    q.push(new TreeNode(-10001));
                }
            }
            if (cur->right) {
                q.push(cur->right);
            } else {
                // Leaf node, don't add null nodes for height = 1
                if (height > 0) {
                    q.push(new TreeNode(10001));
                }
            }
        }
        // std::cout << "2. Traversal: " << traversal << "; height: " << height << std::endl;
    }
    // bfsHelper - A BFS Helper prints straight-forward traversal and also returns height as extra detail
    void bfsHelper(TreeNode* root, std::string& traversal, uint8_t& height) {
        std::queue<TreeNode*> q;
        height = 0;
        if (!root) {
            return;
        }
        if (!root->left && !root->right) {
            traversal.append(std::to_string(root->val));
            return;
        }
        q.push(root);
        q.push(nullptr); // Signifies end of level traversal
        while(!q.empty()) {
            TreeNode* cur = q.front(); q.pop();
            if (!cur) {
                // increment height
                height++;
                if (q.empty()) {
                    // All nodes processed
                    break;
                }
                q.push(nullptr);
                continue;
            } else {
                traversal.append(std::to_string(cur->val));
                traversal.append(",");
            }
            if (cur->left) {
                q.push(cur->left);
            }
            if (cur->right) {
                q.push(cur->right);
            }
        }
        // std::cout << "1. Traversal: " << traversal << "; height: " << height << std::endl;
    }
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if (p == q || !p && !q) {
            return true;
        }
        if (!p || !q) {
            return false;
        }
        std::string p_traversal, q_traversal;
        uint8_t p_height, q_height;
        bfsHelper(p, p_traversal, p_height);
        bfsHelper(q, q_traversal, q_height);
        if (p_traversal != q_traversal || p_height != q_height) {
            return false;
        }
        p_traversal.clear(); q_traversal.clear();
        detailedBfsHelper(p, p_traversal, p_height);
        detailedBfsHelper(q, q_traversal, q_height);
        return (p_traversal == q_traversal);
    }
};

File: solutions/0100-same-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
import aq "github.com/emirpasic/gods/queues/arrayqueue"

func bfsHelper(root *TreeNode) string {
    var bfsOutput string
    queue := aq.New()
    queue.Enqueue(root)
    for !queue.Empty() {
        curEle, _ := queue.Dequeue()
        cur := curEle.(*TreeNode)
        if cur.Val == 10001 {
            bfsOutput = strings.Join([]string{bfsOutput, "nullptr"}, ",")
            curEle = nil // for gc
            continue
        }
        bfsOutput = strings.Join([]string{bfsOutput, strconv.Itoa(cur.Val)}, ",")
        if cur.Left != nil {
            queue.Enqueue(cur.Left)
        } else {
            queue.Enqueue(&TreeNode{
                Val: 10001, Left: nil, Right: nil,
            })
        }
        if cur.Right != nil {
            queue.Enqueue(cur.Right)
        } else {
            queue.Enqueue(&TreeNode{
                Val: 10001, Left: nil, Right: nil,
            })
        }
    }
    return bfsOutput
}

func isSameTree(p *TreeNode, q *TreeNode) bool {
    if p == nil && q == nil {
        return true
    }
    if p == nil || q == nil {
        return false
    }
    pOutput := bfsHelper(p)
    qOutput := bfsHelper(q)
    return pOutput == qOutput
}

